pkg/maglev/maglev.go

```diff
@@ -74,33 +74,58 @@ func getPermutation(backends []string, m uint64) []uint64 {
 	return perm
 }
 
+//BackendPoint is a backend point with weight
+type BackendPoint struct {
+	ID     uint16
+	Weight uint32
+}
+
 // GetLookupTable returns the Maglev lookup table of the size "m" for the given
 // backends. The lookup table contains the indices of the given backends.
-func GetLookupTable(backends []string, m uint64) []int {
-	if len(backends) == 0 {
-		return nil
+func GetLookupTable(backends map[string]*BackendPoint, m uint64, maglevBackendIDsBuffer []uint16) {
+	backendNames := make([]string, 0, len(backends))
+	for name := range backends {
+		backendNames = append(backendNames, name)
 	}
-
-	perm := getPermutation(backends, m)
-	next := make([]int, len(backends))
+	// Maglev algorithm might produce different lookup table for the same
+	// set of backends listed in a different order. To avoid that sort
+	// backends by name, as the names are the same on all nodes (in opposite
+	// to backend IDs which are node-local).
+	sort.Strings(backendNames)
+
+	perm := getPermutation(backendNames, m)
+	next := make([]int, len(backendNames))
 	entry := make([]int, m)
 
 	for j := uint64(0); j < m; j++ {
 		entry[j] = -1
 	}
 
-	l := len(backends)
-
-	for n := uint64(0); n < m; n++ {
-		i := int(n) % l
-		c := perm[i*int(m)+next[i]]
-		for entry[c] >= 0 {
-			next[i] += 1
-			c = perm[i*int(m)+next[i]]
+	runs := uint64(0)
+	for {
+		for i, backendName := range backendNames {
+			//Support weight for backend.
+			//Current implementation assumes that sum of all weights must be more or less the size of hashing ring (by default 65537).
+			//So in your example on of em, 1 VIP and 2 real endpoints. You should have weight 6k and another ~60k (to have 1:10 ratio).
+			//Here (at fb) what we do, is using logic in control plane of balancer which is working like this:
+			//	1.Calculate sum of all real. (Eg 1+10)
+			//	2.Devide hash ring size by this sum. (65537/11 = 5957)
+			//	3.Allocate weight to each by multiple it’s original weight with number from step 2 (1*5957 for real one. 10 * 5957 for real 2)
```

Comment:
Thanks for the extra detail. Is this comment meant to stay in the source or was this just to inform reviewers? It seems useful enough to keep, but we can strip out phrases like "So in your example on of em," and "Here at (fb)".

Reply:
Thanks. You are right, I would keep this comment and strip out referenced phrases.